internal/pkg/runtime/engines/singularity/container_linux.go

```diff
@@ -1339,13 +1341,23 @@ func (c *container) addUserbindsMount(system *mount.System) error {
 
 		sylog.Debugf("Adding %s to mount list\n", src)
 
+		if c.userNS {
+			// test if the source directory is a read-only filesystem by
+			// asking for write permission for this directory, if EROFS error
+			// is returned, we need to skip remount to avoid failure
+			if err := syscall.Access(src, 0x2); err.(syscall.Errno) == syscall.EROFS {
```

Comment:
Since it is running unprivileged, if the source directory is simply owned by root, won't this treat it as a read-only filesystem?  On the other hand, maybe it doesn't matter, because in that case it won't try to remount.  Or will it?  The remount seems to otherwise be always tried when AddBind() below returns err == nil.

For that matter, why do this check here and have a remount flag at all?  Why not do the check down below, just before the AddRemount, only when it is actually needed?

Reply:
In version 2.6.1, the [corresponding code](https://github.com/sylabs/singularity/blob/vault/release-2.6/src/lib/runtime/mounts/userbinds/userbinds.c#L143) only sets read_only when the user requests it and then [skips the remount with a warning](https://github.com/sylabs/singularity/blob/vault/release-2.6/src/lib/runtime/mounts/userbinds/userbinds.c#L143).  It does not, as far as I see, detect anywhere that the source filesystem is read-only and try to copy that.  How's that being detected now?  Even with this patch if the user requests read only it will fail rather than warn.

In a couple of other places in this same source file, when mounting /proc and /sys, it skips the remount whenever c.userNS is set.  Why not do it the same way?  And there are a number of other places where AddRemount is called (e.g. in addHostMount) that do not check c.userNS; won't some of them have this same problem?

Reply:
`EROFS` will take precedence over `EPERM`, that's why `EPERM` is ignored here because it means the directory is owned by another user and it doesn't matter in this case.

Regarding 2.6.1 behavior, it was a lazy approach, IMO when we can we should apply mount flag correctly even with user namespace enabled (users may want to bind a directory in read-only).

I will apply fix for others. Thanks !